service/history/workflow/mutable_state_impl.go

```diff
@@ -4603,6 +4654,26 @@ func (ms *MutableStateImpl) GenerateMigrationTasks() ([]tasks.Task, int64, error
 	return ms.taskGenerator.GenerateMigrationTasks()
 }
 
+func (ms *MutableStateImpl) processCallbacks() error {
```

Comment:
Ideally all task generation logic should be in mutableStateTaskGenerator. GenerateWorkflowCloseTasks() in this case. Is it possible to move it there?

~~Also if we do that, then it automatically make the task refresher work.~~
We also need to make task refresher work with callback tasks.

Reply:
Hmm... I'll look into this.

I wanted the tasks to be generated as part of the callback state transition, but this may not be possible if we do state based replication for callbacks. We'll figure that out when we implement replication.

@MichaelSnowden heads up.

Reply:
Okay, I see what you mean.
I really don't like this approach but I'll keep it for consistency with the rest of the code.